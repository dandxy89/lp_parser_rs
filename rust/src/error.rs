//! Comprehensive error types for LP parsing operations.
//!
//! This module provides a structured error hierarchy with detailed context
//! information for debugging and error handling in LP file parsing.

use thiserror::Error;

/// Comprehensive error type for LP parsing operations.
///
/// This error type provides detailed context about parsing failures,
/// including location information and specific error conditions.
#[derive(Error, Debug, Clone, PartialEq, Eq)]
pub enum LpParseError {
    /// Invalid or malformed constraint syntax
    #[error("Invalid constraint syntax at position {position}: {context}")]
    ConstraintSyntax { position: usize, context: String },

    /// Invalid or malformed objective syntax
    #[error("Invalid objective syntax at position {position}: {context}")]
    ObjectiveSyntax { position: usize, context: String },

    /// Unknown or invalid variable type specification
    #[error("Unknown variable type '{var_type}' for variable '{variable}'")]
    UnknownVariableType { variable: String, var_type: String },

    /// Reference to an undefined variable
    #[error("Undefined variable '{variable}' referenced in {context}")]
    UndefinedVariable { variable: String, context: String },

    /// Duplicate definition of a component
    #[error("Duplicate {component_type} '{name}' defined")]
    DuplicateDefinition { component_type: String, name: String },

    /// Invalid numerical value or format
    #[error("Invalid number format '{value}' at position {position}")]
    InvalidNumber { value: String, position: usize },

    /// Missing required section in LP file
    #[error("Missing required section: {section}")]
    MissingSection { section: String },

    /// Invalid bound specification
    #[error("Invalid bounds for variable '{variable}': {details}")]
    InvalidBounds { variable: String, details: String },

    /// Invalid SOS constraint specification
    #[error("Invalid SOS constraint '{constraint}': {details}")]
    InvalidSosConstraint { constraint: String, details: String },

    /// Validation error for logical consistency
    #[error("Validation error: {message}")]
    ValidationError { message: String },

    /// Generic parsing error with context
    #[error("Parse error at position {position}: {message}")]
    ParseError { position: usize, message: String },

    /// File I/O related errors
    #[error("File I/O error: {message}")]
    IoError { message: String },

    /// Internal parser state errors
    #[error("Internal parser error: {message}")]
    InternalError { message: String },
}

impl LpParseError {
    /// Create a new constraint syntax error
    pub fn constraint_syntax(position: usize, context: impl Into<String>) -> Self {
        Self::ConstraintSyntax { position, context: context.into() }
    }

    /// Create a new objective syntax error
    pub fn objective_syntax(position: usize, context: impl Into<String>) -> Self {
        Self::ObjectiveSyntax { position, context: context.into() }
    }

    /// Create a new unknown variable type error
    pub fn unknown_variable_type(variable: impl Into<String>, var_type: impl Into<String>) -> Self {
        Self::UnknownVariableType { variable: variable.into(), var_type: var_type.into() }
    }

    /// Create a new undefined variable error
    pub fn undefined_variable(variable: impl Into<String>, context: impl Into<String>) -> Self {
        Self::UndefinedVariable { variable: variable.into(), context: context.into() }
    }

    /// Create a new duplicate definition error
    pub fn duplicate_definition(component_type: impl Into<String>, name: impl Into<String>) -> Self {
        Self::DuplicateDefinition { component_type: component_type.into(), name: name.into() }
    }

    /// Create a new invalid number error
    pub fn invalid_number(value: impl Into<String>, position: usize) -> Self {
        Self::InvalidNumber { value: value.into(), position }
    }

    /// Create a new missing section error
    pub fn missing_section(section: impl Into<String>) -> Self {
        Self::MissingSection { section: section.into() }
    }

    /// Create a new invalid bounds error
    pub fn invalid_bounds(variable: impl Into<String>, details: impl Into<String>) -> Self {
        Self::InvalidBounds { variable: variable.into(), details: details.into() }
    }

    /// Create a new invalid SOS constraint error
    pub fn invalid_sos_constraint(constraint: impl Into<String>, details: impl Into<String>) -> Self {
        Self::InvalidSosConstraint { constraint: constraint.into(), details: details.into() }
    }

    /// Create a new validation error
    pub fn validation_error(message: impl Into<String>) -> Self {
        Self::ValidationError { message: message.into() }
    }

    /// Create a new parse error
    pub fn parse_error(position: usize, message: impl Into<String>) -> Self {
        Self::ParseError { position, message: message.into() }
    }

    /// Create a new I/O error
    pub fn io_error(message: impl Into<String>) -> Self {
        Self::IoError { message: message.into() }
    }

    /// Create a new internal error
    pub fn internal_error(message: impl Into<String>) -> Self {
        Self::InternalError { message: message.into() }
    }
}

/// Convert from nom parsing errors to our custom error type
impl<'a> From<nom::Err<nom::error::Error<&'a str>>> for LpParseError {
    fn from(err: nom::Err<nom::error::Error<&'a str>>) -> Self {
        match err {
            nom::Err::Incomplete(_) => Self::parse_error(0, "Incomplete input"),
            nom::Err::Error(e) | nom::Err::Failure(e) => Self::parse_error(e.input.as_ptr() as usize, format!("Parse error: {:?}", e.code)),
        }
    }
}

/// Convert from standard I/O errors
impl From<std::io::Error> for LpParseError {
    fn from(err: std::io::Error) -> Self {
        Self::io_error(err.to_string())
    }
}

/// Convert from boxed errors (used by CSV module)
impl From<Box<dyn std::error::Error>> for LpParseError {
    fn from(err: Box<dyn std::error::Error>) -> Self {
        Self::io_error(err.to_string())
    }
}

/// Result type alias for LP parsing operations
pub type LpResult<T> = Result<T, LpParseError>;

/// Context extension trait for adding location information to errors
pub trait ErrorContext<T> {
    /// Add position context to an error
    fn with_position(self, position: usize) -> LpResult<T>;

    /// Add general context to an error
    fn with_context(self, context: &str) -> LpResult<T>;
}

impl<T> ErrorContext<T> for Result<T, LpParseError> {
    fn with_position(self, position: usize) -> Self {
        self.map_err(|mut err| {
            if let LpParseError::ParseError { position: ref mut pos, .. } = &mut err {
                *pos = position;
            }
            err
        })
    }

    fn with_context(self, context: &str) -> Self {
        self.map_err(|err| match err {
            LpParseError::ParseError { position, message } => LpParseError::parse_error(position, format!("{context}: {message}")),
            other => other,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_error_creation() {
        let err = LpParseError::constraint_syntax(42, "missing operator");
        assert_eq!(err.to_string(), "Invalid constraint syntax at position 42: missing operator");
    }

    #[test]
    fn test_error_context() {
        let result: LpResult<()> = Err(LpParseError::parse_error(10, "test error"));
        let with_context = result.with_context("parsing constraint");

        assert!(with_context.is_err());
        assert!(with_context.unwrap_err().to_string().contains("parsing constraint"));
    }

    #[test]
    fn test_io_error_conversion() {
        let io_err = std::io::Error::new(std::io::ErrorKind::NotFound, "file not found");
        let lp_err: LpParseError = io_err.into();

        match lp_err {
            LpParseError::IoError { message } => assert!(message.contains("file not found")),
            _ => panic!("Expected IoError"),
        }
    }
}
