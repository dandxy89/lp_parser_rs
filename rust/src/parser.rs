//! File I/O and content parsing utilities.
//!
//! This module provides functionality for reading LP files from disk
//! and preparing their contents for parsing.
//!

use std::fs::File;
use std::io::{BufReader, Read as _};
use std::path::Path;

use crate::error::{LpParseError, LpResult};

#[inline]
/// Parses the contents of a file at the given path into a string.
///
/// # Arguments
///
/// * `path` - A reference to a `Path` that represents the file path to be parsed.
///
/// # Returns
///
/// A `Result` containing the file contents as a `String` if successful, or an `LpParseError`
/// if the file cannot be opened or read.
///
/// # Errors
///
/// Returns an `LpParseError::IoError` if the file cannot be opened or read.
pub fn parse_file(path: &Path) -> LpResult<String> {
    let file = File::open(path).map_err(|e| LpParseError::io_error(format!("Failed to open file '{}': {}", path.display(), e)))?;

    let mut buf_reader = BufReader::new(file);
    let mut contents = String::new();

    buf_reader
        .read_to_string(&mut contents)
        .map_err(|e| LpParseError::io_error(format!("Failed to read file '{}': {}", path.display(), e)))?;

    Ok(contents)
}
