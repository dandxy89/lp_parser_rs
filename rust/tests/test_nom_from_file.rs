//! This Rust code snippet defines a function to read LP (Linear Programming) files from a resources
//! directory and parse them using the `lp_parser_rs` library.
//!
//! It includes several test functions and a macro to generate tests for various LP files, ensuring they can be parsed correctly.
//!

use std::error::Error;
use std::path::PathBuf;

use lp_parser_rs::parser::parse_file;
use lp_parser_rs::problem::LpProblem;

fn read_file_from_resources(file_name: &str) -> Result<String, Box<dyn Error + 'static>> {
    let mut file_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    file_path.push(format!("resources/{file_name}"));
    let contents = parse_file(&file_path)?;

    Ok(contents)
}

#[test]
fn invalid() {
    let input = read_file_from_resources("invalid.lp").expect("failed to read file from resources");
    assert!(LpProblem::parse(&input).is_err());
}

#[test]
fn nom_fit2d() {
    let input = read_file_from_resources("fit2d.lp").expect("failed to read file from resources");
    let parsed = LpProblem::parse(&input).expect("failed to parse LPProblem");
    assert_eq!(parsed.objective_count(), 1);
    assert_eq!(parsed.constraint_count(), 25);
    assert_eq!(parsed.variable_count(), 10500);

    #[cfg(feature = "serde")]
    insta::assert_yaml_snapshot!(parsed, {
        ".variables" => insta::sorted_redaction(),
        ".objectives" => insta::sorted_redaction(),
        ".constraints" => insta::sorted_redaction(),
    });
}

#[ignore]
#[test]
/// A test used my contributors for assisting with issues
fn nom_debug() {
    let input = read_file_from_resources("infile_comments.lp").expect("failed to read file from resources");
    assert!(LpProblem::parse(&input).is_ok());
}

#[macro_export]
macro_rules! generate_test {
    ($test_name:ident, $file:expr) => {
        #[test]
        #[allow(unused_variables)]
        fn $test_name() {
            let input = read_file_from_resources($file).unwrap();
            let parsed = LpProblem::parse(&input).expect("failed to parse LPProblem");

            #[cfg(feature = "serde")]
            insta::assert_yaml_snapshot!(parsed, {
                ".variables" => insta::sorted_redaction(),
                ".objectives" => insta::sorted_redaction(),
                ".constraints" => insta::sorted_redaction(),
            });
        }
    };
}

generate_test!(pulp, "pulp.lp");
generate_test!(pulp2, "pulp2.lp");
generate_test!(limbo, "limbo.lp");
// generate_test!(semi_continuous, "semi_continuous.lp");
generate_test!(sos, "sos.lp");
generate_test!(test, "test.lp");
generate_test!(test2, "test2.lp");
generate_test!(empty_bounds, "empty_bounds.lp");
generate_test!(blank_lines, "blank_lines.lp");
// generate_test!(optional_labels, "optional_labels.lp");
generate_test!(infile_comments, "infile_comments.lp");
generate_test!(infile_comments2, "infile_comments2.lp");
generate_test!(missing_signs, "missing_signs.lp");
generate_test!(scientific_notation_2, "scientific_notation_2.lp");
generate_test!(output, "output.lp");
generate_test!(output2_1, "output2_1.lp");
generate_test!(output2_2, "output2_2.lp");
generate_test!(output2_3, "output2_3.lp");
generate_test!(output2_4, "output2_4.lp");
generate_test!(complex_names, "complex_names.lp");

// Test files from various open source projects on Github

// Mosek: <https://docs.mosek.com/latest/capi/lp-format.html>
generate_test!(scientific_notation, "scientific_notation.lp");
generate_test!(mosek, "mosek.lp");
generate_test!(mosek_bounds, "mosek_bounds.lp");
generate_test!(lol, "lol.lp");
generate_test!(milo1, "milo1.lp");

// From <https://github.com/asbestian/jplex>
generate_test!(obj3_2cons, "3obj_2cons.lp");

// generate_test!(no_end_section, "no_end_section.lp");
generate_test!(obj_2cons_only_binary_vars, "2obj_2cons_only_binary_vars.lp");
generate_test!(obj_2cons_all_variable_types, "2obj_2cons_all_variable_types.lp");
generate_test!(obj_1cons_all_variables_with_bounds, "1obj_1cons_all_variables_with_bounds.lp");
generate_test!(afiro, "afiro.lp");
generate_test!(afiro_ext, "afiro_ext.lp");
generate_test!(boeing1, "boeing1.lp");
generate_test!(boeing2, "boeing2.lp");
generate_test!(fit1d, "fit1d.lp");
generate_test!(kb2, "kb2.lp");
generate_test!(sc50a, "sc50a.lp");

// From <https://github.com/odow/LPWriter.jl>
generate_test!(model2, "model2.lp");
#[test]
fn corrupt() {
    let input = read_file_from_resources("corrupt.lp").expect("failed to read file from resources");
    assert!(LpProblem::parse(&input).is_err());
}

// From <https://github.com/brymck/lp-parse/tree>
generate_test!(model, "model.lp");

// From <https://github.com/josephcslater/JupyterExamples>
generate_test!(sudoku, "sudoku.lp");

// From <https://github.com/afaf-taik/vehicularFL>
generate_test!(wbm, "wbm.lp");

// From <https://github.com/IBMDecisionOptimization/cplexrunonwml>
generate_test!(diet, "diet.lp");

// From <https://github.com/claudiosa/CCS>
// generate_test!(output_cplex_2, "output_cplex_2.lp");

// From <https://github.com/coin-or/pulp>
generate_test!(american_steel_problem, "AmericanSteelProblem.lp");
generate_test!(beer_distribution_problem, "BeerDistributionProblem.lp");
generate_test!(computer_plant_problem, "ComputerPlantProblem.lp");
generate_test!(whiskas_model_2, "WhiskasModel2.lp");

// From <https://lpsolve.sourceforge.net/5.0/CPLEX-format.htm>
generate_test!(cplex, "cplex.lp");
