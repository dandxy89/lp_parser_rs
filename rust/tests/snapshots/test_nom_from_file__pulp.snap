---
source: rust/tests/test_nom_from_file.rs
expression: parsed
---
name: " copynumber "
sense: Minimize
objectives:
  OBJ:
    name: OBJ
    coefficients:
      - name: PenaltyTree_48
        value: 1
constraints:
  _C1:
    type: Standard
    name: _C1
    coefficients:
      - name: A_0
        value: -1
      - name: A_3
        value: 1
      - name: over_0
        value: -1
    operator: LTE
    rhs: 0
  _C10:
    type: Standard
    name: _C10
    coefficients:
      - name: B_3
        value: -1
      - name: B_7
        value: 1
      - name: under_9
        value: 1
    operator: GTE
    rhs: 0
  _C11:
    type: Standard
    name: _C11
    coefficients:
      - name: D_0
        value: -1
      - name: D_3
        value: 1
      - name: over_10
        value: -1
    operator: LTE
    rhs: 0
  _C12:
    type: Standard
    name: _C12
    coefficients:
      - name: D_0
        value: -1
      - name: D_3
        value: 1
      - name: under_11
        value: 1
    operator: GTE
    rhs: 0
  _C13:
    type: Standard
    name: _C13
    coefficients:
      - name: N_0
        value: -1
      - name: N_3
        value: 1
      - name: over_12
        value: -1
    operator: LTE
    rhs: 0
  _C14:
    type: Standard
    name: _C14
    coefficients:
      - name: N_0
        value: -1
      - name: N_3
        value: 1
      - name: under_13
        value: 1
    operator: GTE
    rhs: 0
  _C15:
    type: Standard
    name: _C15
    coefficients:
      - name: N_3
        value: -1
      - name: N_7
        value: 1
      - name: over_14
        value: -1
    operator: LTE
    rhs: 0
  _C16:
    type: Standard
    name: _C16
    coefficients:
      - name: N_3
        value: -1
      - name: N_7
        value: 1
      - name: under_15
        value: 1
    operator: GTE
    rhs: 0
  _C17:
    type: Standard
    name: _C17
    coefficients:
      - name: A_7
        value: 1
      - name: B_7
        value: -1
      - name: over_16
        value: -1
    operator: LTE
    rhs: 0
  _C18:
    type: Standard
    name: _C18
    coefficients:
      - name: A_7
        value: 1
      - name: B_7
        value: -1
      - name: under_17
        value: 1
    operator: GTE
    rhs: 0
  _C19:
    type: Standard
    name: _C19
    coefficients:
      - name: A_7
        value: -1
      - name: N_7
        value: 1
      - name: over_18
        value: -1
    operator: LTE
    rhs: 0
  _C2:
    type: Standard
    name: _C2
    coefficients:
      - name: A_0
        value: -1
      - name: A_3
        value: 1
      - name: under_1
        value: 1
    operator: GTE
    rhs: 0
  _C20:
    type: Standard
    name: _C20
    coefficients:
      - name: A_7
        value: -1
      - name: N_7
        value: 1
      - name: under_19
        value: 1
    operator: GTE
    rhs: 0
  _C21:
    type: Standard
    name: _C21
    coefficients:
      - name: C_7
        value: 1
      - name: N_7
        value: -1
      - name: over_20
        value: -1
    operator: LTE
    rhs: 0
  _C22:
    type: Standard
    name: _C22
    coefficients:
      - name: C_7
        value: 1
      - name: N_7
        value: -1
      - name: under_21
        value: 1
    operator: GTE
    rhs: 0
  _C23:
    type: Standard
    name: _C23
    coefficients:
      - name: B_7
        value: 1
      - name: C_7
        value: -1
      - name: over_22
        value: -1
    operator: LTE
    rhs: 0
  _C24:
    type: Standard
    name: _C24
    coefficients:
      - name: B_7
        value: 1
      - name: C_7
        value: -1
      - name: under_23
        value: 1
    operator: GTE
    rhs: 0
  _C25:
    type: Standard
    name: _C25
    coefficients:
      - name: B_3
        value: 1
      - name: N_3
        value: -1
      - name: over_24
        value: -1
    operator: LTE
    rhs: 0
  _C26:
    type: Standard
    name: _C26
    coefficients:
      - name: B_3
        value: 1
      - name: N_3
        value: -1
      - name: under_25
        value: 1
    operator: GTE
    rhs: 0
  _C27:
    type: Standard
    name: _C27
    coefficients:
      - name: A_3
        value: 1
      - name: B_3
        value: -1
      - name: over_26
        value: -1
    operator: LTE
    rhs: 0
  _C28:
    type: Standard
    name: _C28
    coefficients:
      - name: A_3
        value: 1
      - name: B_3
        value: -1
      - name: under_27
        value: 1
    operator: GTE
    rhs: 0
  _C29:
    type: Standard
    name: _C29
    coefficients:
      - name: A_3
        value: -1
      - name: N_3
        value: 1
      - name: over_28
        value: -1
    operator: LTE
    rhs: 0
  _C3:
    type: Standard
    name: _C3
    coefficients:
      - name: A_3
        value: -1
      - name: A_7
        value: 1
      - name: over_2
        value: -1
    operator: LTE
    rhs: 0
  _C30:
    type: Standard
    name: _C30
    coefficients:
      - name: A_3
        value: -1
      - name: N_3
        value: 1
      - name: under_29
        value: 1
    operator: GTE
    rhs: 0
  _C31:
    type: Standard
    name: _C31
    coefficients:
      - name: A_0
        value: 1
      - name: N_0
        value: -1
      - name: over_30
        value: -1
    operator: LTE
    rhs: 0
  _C32:
    type: Standard
    name: _C32
    coefficients:
      - name: A_0
        value: 1
      - name: N_0
        value: -1
      - name: under_31
        value: 1
    operator: GTE
    rhs: 0
  _C33:
    type: Standard
    name: _C33
    coefficients:
      - name: A_0
        value: -1
      - name: D_0
        value: 1
      - name: over_32
        value: -1
    operator: LTE
    rhs: 0
  _C34:
    type: Standard
    name: _C34
    coefficients:
      - name: A_0
        value: -1
      - name: D_0
        value: 1
      - name: under_33
        value: 1
    operator: GTE
    rhs: 0
  _C35:
    type: Standard
    name: _C35
    coefficients:
      - name: B_0
        value: 1
      - name: D_0
        value: -1
      - name: over_34
        value: -1
    operator: LTE
    rhs: 0
  _C36:
    type: Standard
    name: _C36
    coefficients:
      - name: B_0
        value: 1
      - name: D_0
        value: -1
      - name: under_35
        value: 1
    operator: GTE
    rhs: 0
  _C37:
    type: Standard
    name: _C37
    coefficients:
      - name: B_0
        value: -1
      - name: N_0
        value: 1
      - name: over_36
        value: -1
    operator: LTE
    rhs: 0
  _C38:
    type: Standard
    name: _C38
    coefficients:
      - name: B_0
        value: -1
      - name: N_0
        value: 1
      - name: under_37
        value: 1
    operator: GTE
    rhs: 0
  _C39:
    type: Standard
    name: _C39
    coefficients:
      - name: D_3
        value: 1
      - name: over_38
        value: -1
    operator: LTE
    rhs: 1.24660824696
  _C4:
    type: Standard
    name: _C4
    coefficients:
      - name: A_3
        value: -1
      - name: A_7
        value: 1
      - name: under_3
        value: 1
    operator: GTE
    rhs: 0
  _C40:
    type: Standard
    name: _C40
    coefficients:
      - name: D_3
        value: 1
      - name: under_39
        value: 1
    operator: GTE
    rhs: 1.24660824696
  _C41:
    type: Standard
    name: _C41
    coefficients:
      - name: A_7
        value: 1
      - name: B_7
        value: 1
      - name: C_7
        value: 1
      - name: N_7
        value: 1
      - name: over_40
        value: -1
    operator: LTE
    rhs: 9.80665154278
  _C42:
    type: Standard
    name: _C42
    coefficients:
      - name: A_7
        value: 1
      - name: B_7
        value: 1
      - name: C_7
        value: 1
      - name: N_7
        value: 1
      - name: under_41
        value: 1
    operator: GTE
    rhs: 9.80665154278
  _C43:
    type: Standard
    name: _C43
    coefficients:
      - name: C_0
        value: 1
      - name: over_42
        value: -1
    operator: LTE
    rhs: 2.06580795211
  _C44:
    type: Standard
    name: _C44
    coefficients:
      - name: C_0
        value: 1
      - name: under_43
        value: 1
    operator: GTE
    rhs: 2.06580795211
  _C45:
    type: Standard
    name: _C45
    coefficients:
      - name: A_3
        value: 1
      - name: B_3
        value: 1
      - name: N_3
        value: 1
      - name: over_44
        value: -1
    operator: LTE
    rhs: 7.1056670077
  _C46:
    type: Standard
    name: _C46
    coefficients:
      - name: A_3
        value: 1
      - name: B_3
        value: 1
      - name: N_3
        value: 1
      - name: under_45
        value: 1
    operator: GTE
    rhs: 7.1056670077
  _C47:
    type: Standard
    name: _C47
    coefficients:
      - name: A_0
        value: 1
      - name: B_0
        value: 1
      - name: D_0
        value: 1
      - name: N_0
        value: 1
      - name: over_46
        value: -1
    operator: LTE
    rhs: 7.89518556411
  _C48:
    type: Standard
    name: _C48
    coefficients:
      - name: A_0
        value: 1
      - name: B_0
        value: 1
      - name: D_0
        value: 1
      - name: N_0
        value: 1
      - name: under_47
        value: 1
    operator: GTE
    rhs: 7.89518556411
  _C49:
    type: Standard
    name: _C49
    coefficients:
      - name: PenaltyTree_48
        value: -1
      - name: over_0
        value: 0.5
      - name: over_10
        value: 0.5
      - name: over_12
        value: 0.5
      - name: over_14
        value: 0.5
      - name: over_16
        value: 0.02
      - name: over_18
        value: 0.02
      - name: over_2
        value: 0.5
      - name: over_20
        value: 0.02
      - name: over_22
        value: 0.02
      - name: over_24
        value: 0.02
      - name: over_26
        value: 0.02
      - name: over_28
        value: 0.02
      - name: over_30
        value: 0.02
      - name: over_32
        value: 0.02
      - name: over_34
        value: 0.02
      - name: over_36
        value: 0.02
      - name: over_38
        value: 4
      - name: over_4
        value: 0.5
      - name: over_40
        value: 4
      - name: over_42
        value: 4
      - name: over_44
        value: 4
      - name: over_46
        value: 4
      - name: over_6
        value: 0.5
      - name: over_8
        value: 0.5
      - name: under_1
        value: 0.5
      - name: under_11
        value: 0.5
      - name: under_13
        value: 0.5
      - name: under_15
        value: 0.5
      - name: under_17
        value: 0.02
      - name: under_19
        value: 0.02
      - name: under_21
        value: 0.02
      - name: under_23
        value: 0.02
      - name: under_25
        value: 0.02
      - name: under_27
        value: 0.02
      - name: under_29
        value: 0.02
      - name: under_3
        value: 0.5
      - name: under_31
        value: 0.02
      - name: under_33
        value: 0.02
      - name: under_35
        value: 0.02
      - name: under_37
        value: 0.02
      - name: under_39
        value: 4
      - name: under_41
        value: 4
      - name: under_43
        value: 4
      - name: under_45
        value: 4
      - name: under_47
        value: 4
      - name: under_5
        value: 0.5
      - name: under_7
        value: 0.5
      - name: under_9
        value: 0.5
    operator: EQ
    rhs: 0
  _C5:
    type: Standard
    name: _C5
    coefficients:
      - name: C_0
        value: -1
      - name: C_7
        value: 1
      - name: over_4
        value: -1
    operator: LTE
    rhs: 0
  _C6:
    type: Standard
    name: _C6
    coefficients:
      - name: C_0
        value: -1
      - name: C_7
        value: 1
      - name: under_5
        value: 1
    operator: GTE
    rhs: 0
  _C7:
    type: Standard
    name: _C7
    coefficients:
      - name: B_0
        value: -1
      - name: B_3
        value: 1
      - name: over_6
        value: -1
    operator: LTE
    rhs: 0
  _C8:
    type: Standard
    name: _C8
    coefficients:
      - name: B_0
        value: -1
      - name: B_3
        value: 1
      - name: under_7
        value: 1
    operator: GTE
    rhs: 0
  _C9:
    type: Standard
    name: _C9
    coefficients:
      - name: B_3
        value: -1
      - name: B_7
        value: 1
      - name: over_8
        value: -1
    operator: LTE
    rhs: 0
variables:
  A_0:
    name: A_0
    var_type:
      DoubleBound:
        - 0
        - 4
  A_3:
    name: A_3
    var_type:
      DoubleBound:
        - 0
        - 4
  A_7:
    name: A_7
    var_type:
      DoubleBound:
        - 0
        - 4
  B_0:
    name: B_0
    var_type:
      DoubleBound:
        - 0
        - 4
  B_3:
    name: B_3
    var_type:
      DoubleBound:
        - 0
        - 4
  B_7:
    name: B_7
    var_type:
      DoubleBound:
        - 0
        - 4
  C_0:
    name: C_0
    var_type:
      DoubleBound:
        - 0
        - 4
  C_7:
    name: C_7
    var_type:
      DoubleBound:
        - 0
        - 4
  D_0:
    name: D_0
    var_type:
      DoubleBound:
        - 0
        - 4
  D_3:
    name: D_3
    var_type:
      DoubleBound:
        - 0
        - 4
  N_0:
    name: N_0
    var_type:
      DoubleBound:
        - 0
        - 4
  N_3:
    name: N_3
    var_type:
      DoubleBound:
        - 0
        - 4
  N_7:
    name: N_7
    var_type:
      DoubleBound:
        - 0
        - 4
  PenaltyTree_48:
    name: PenaltyTree_48
    var_type: Free
  over_0:
    name: over_0
    var_type: Free
  over_10:
    name: over_10
    var_type: Free
  over_12:
    name: over_12
    var_type: Free
  over_14:
    name: over_14
    var_type: Free
  over_16:
    name: over_16
    var_type: Free
  over_18:
    name: over_18
    var_type: Free
  over_2:
    name: over_2
    var_type: Free
  over_20:
    name: over_20
    var_type: Free
  over_22:
    name: over_22
    var_type: Free
  over_24:
    name: over_24
    var_type: Free
  over_26:
    name: over_26
    var_type: Free
  over_28:
    name: over_28
    var_type: Free
  over_30:
    name: over_30
    var_type: Free
  over_32:
    name: over_32
    var_type: Free
  over_34:
    name: over_34
    var_type: Free
  over_36:
    name: over_36
    var_type: Free
  over_38:
    name: over_38
    var_type: Free
  over_4:
    name: over_4
    var_type: Free
  over_40:
    name: over_40
    var_type: Free
  over_42:
    name: over_42
    var_type: Free
  over_44:
    name: over_44
    var_type: Free
  over_46:
    name: over_46
    var_type: Free
  over_6:
    name: over_6
    var_type: Free
  over_8:
    name: over_8
    var_type: Free
  under_1:
    name: under_1
    var_type: Free
  under_11:
    name: under_11
    var_type: Free
  under_13:
    name: under_13
    var_type: Free
  under_15:
    name: under_15
    var_type: Free
  under_17:
    name: under_17
    var_type: Free
  under_19:
    name: under_19
    var_type: Free
  under_21:
    name: under_21
    var_type: Free
  under_23:
    name: under_23
    var_type: Free
  under_25:
    name: under_25
    var_type: Free
  under_27:
    name: under_27
    var_type: Free
  under_29:
    name: under_29
    var_type: Free
  under_3:
    name: under_3
    var_type: Free
  under_31:
    name: under_31
    var_type: Free
  under_33:
    name: under_33
    var_type: Free
  under_35:
    name: under_35
    var_type: Free
  under_37:
    name: under_37
    var_type: Free
  under_39:
    name: under_39
    var_type: Free
  under_41:
    name: under_41
    var_type: Free
  under_43:
    name: under_43
    var_type: Free
  under_45:
    name: under_45
    var_type: Free
  under_47:
    name: under_47
    var_type: Free
  under_5:
    name: under_5
    var_type: Free
  under_7:
    name: under_7
    var_type: Free
  under_9:
    name: under_9
    var_type: Free
